Prevent self-trading

sporejack opened the pull request:
# Motivation
Traders should not be allowed to trade with themselves.

# Changes
 - Simplify unit test suite
 - Add explicit guard against self-matching in `Book::submit`
 - Add associated unit test for this case

sporejack committed:
Simplify unit test setup

This commit greatly simplifies the setup phase of the unit test suite
targeting the `book` module.

sporejack committed:
Add guard against self-matching and associated test

This commit adds an explicit guard against matching with a trader's own
order.

raymogg commented on src/book_tests.rs:
```diff
@@ -303,12 +224,37 @@ pub async fn test_deep_buy() {
     assert_eq!(ask_length, 3);
 }
 
+#[tokio::test]
```
Would be worth testing the branch where the last order on the book is the users own order too

sporejack replied:
Consult 4bd574e.

sporejack committed:
Add test case for low-volume self-matching

This commit adds a unit test case covering the scenario where a trader
attempts to trade with themselves *and is the only one on the opposite
side of the book*.

raymogg commented on src/book.rs:
```diff
@@ -189,6 +189,17 @@ impl Book {
         // match until either the order is fully matched, or all valid orders to match with are done.
         while !done {
+            /* prevent self-matching */
+            if orders_queue.front().is_some()
```
Does `unwrap()` pop the order from the queue? 

If it is, then we don't put it back on the queue after the continue. But if it isn't popping, then how does the call on line 204 to `orders_queue.pop_front()` not pick up this order?

The tests pass so I'm obviously missing something here but just wanted to check my understanding

sporejack replied:
@raymogg [`front`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.front) simply provides an *immutable* reference to the element at the front of the deque, or `None` if the deque is empty. All [`unwrap`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap) does is retrieve the element within the `Option` returned by `front` and panics if the option is `None`.

To avoid a panic, the first term of the conditional bounds checks the length of the deque, which, due to short-circuit evaluation of boolean expressions, will skip the `unwrap` call.
